dali/imgcodec/decoders/nvjpeg.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2019-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/imgcodec/decoders/nvjpeg.h"
+#include "dali/imgcodec/decoders/nvjpeg/nvjpeg_helper.h"
+#include "dali/imgcodec/decoders/nvjpeg/nvjpeg_memory.h"
+#include "dali/imgcodec/decoders/nvjpeg/permute_layout.h"
+
+namespace dali {
+namespace imgcodec {
+
+NvJpegDecoderInstance::NvJpegDecoderInstance(int device_id, ThreadPool *tp)
+: BatchParallelDecoderImpl(device_id, tp),
```

Comment:
Using `BatchParallelDecoderImpl` wouldn't work when we introduce the HW decoder implementation. Because we will have to split the batch and run some samples in the HW decoder and the rest one-by-one in the thread pool. I think it'd be good to keep in mind in this original design, to avoid having to rewrite later.

Reply:
:+1: